prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/ConstantResolver.kt

```diff
@@ -0,0 +1,135 @@
+package gov.cdc.prime.router.fhirengine.translation.hl7.utils
+
+import gov.cdc.prime.router.fhirengine.translation.hl7.HL7ConversionException
+import org.apache.commons.lang3.math.NumberUtils
+import org.apache.commons.text.StringSubstitutor
+import org.apache.commons.text.lookup.StringLookup
+import org.apache.logging.log4j.kotlin.Logging
+import org.hl7.fhir.exceptions.PathEngineException
+import org.hl7.fhir.r4.model.Base
+import org.hl7.fhir.r4.model.IntegerType
+import org.hl7.fhir.r4.model.StringType
+import org.hl7.fhir.r4.model.TypeDetails
+import org.hl7.fhir.r4.model.ValueSet
+import org.hl7.fhir.r4.utils.FHIRPathEngine
+
+/**
+ * Context used for resolving [constants].
+ */
+data class CustomContext(val constants: MutableMap<String, String> = mutableMapOf()) {
+    companion object {
+        /**
+         * Add [constants] to a context.
+         * @return a new context with the [constants] added
+         */
+        fun addConstants(constants: Map<String, String>, previousContext: CustomContext?): CustomContext? {
+            return if (constants.isEmpty()) previousContext
+            else {
+                val newContext = previousContext?.copy() ?: CustomContext()
+                constants.forEach { newContext.constants[it.key] = it.value }
+                newContext
+            }
+        }
+
+        /**
+         * Add constant with [key] and [value] to a context.
+         * @return a new context with the constant added
+         */
+        fun addConstant(key: String, value: String, previousContext: CustomContext?): CustomContext? {
+            return addConstants(mapOf(key to value), previousContext)
+        }
+    }
+}
+
+/**
+ * String substitution for constants.
+ */
+object ConstantSubstitutor {
+    /**
+     * The resolver.  Uses %{} instead of the default to match the FHIR path use of %.
+     */
+    private val constantResolver = StringSubstitutor().setVariablePrefix("%{").setEscapeChar('%')
+
+    /**
+     * Replace the constants in a given [inputText] using the [context].
+     * @return the resolved string
+     */
+    fun replace(inputText: String, context: CustomContext?): String {
+        return constantResolver.setVariableResolver(StringCustomResolver(context)).replace(inputText)
+    }
+
+    /**
+     * Custom resolver for the [ConstantSubstitutor] that uses the [context] to resolve the constants.
+     */
+    internal class StringCustomResolver(val context: CustomContext?) : StringLookup, Logging {
+        override fun lookup(key: String?): String {
+            require(!key.isNullOrBlank())
+            if (context == null || !context.constants.contains(key))
+                throw HL7ConversionException("Constant '$key' cannot be resolved in string")
```

Comment:
Should this say '... cannot be resolved into a String'?

Reply:
I changed this text to read differently.